Monads.NET/Maybe.cs

```diff
@@ -98,10 +99,44 @@ public static TInput IfDo<TInput>(this TInput o, Func<TInput, bool> evaluator, A
         public static TResult With<TKey, TResult>(this IDictionary<TKey, TResult> o, TKey key)
             where TResult : class
         {
-            return (o == null || !o.ContainsKey(key)) ? null : o[key];
+            TResult value;
+            return (o == null || !o.TryGetValue(key, out value)) ? null : value;
         }
 
         /// <summary>
+        /// Allows for a null safe accessing of an item
+        /// </summary>
+        /// <typeparam name="TInput">Reference type of object being extended</typeparam>
+        /// <typeparam name="TResult">Reference type of object being returned</typeparam>
+        /// <param name="o">instance of object being extended</param>
+        /// <param name="evaluator">function that acts on object to return a result</param>
+        /// <returns>null if object is null or an instance of TResult</returns>
+        public static Task<TResult> WithAsync<TInput, TResult>(this TInput o, Func<TInput, Task<TResult>> evaluator)
+            where TResult : class
+            where TInput : class
+        {
+            return (o == null) ? TaskEx.FromResult<TResult>(null) : evaluator(o);
```

Comment:
Your project is .Net 4.0 Client profile
Task.FromResul appeared in .net 4.5

additinaly await keyword is available only in 4.5 (I can't install Async CTP on vs 2010, my installation is broken, and vs2012 supports it only in 4.5)

plus using await creates overhead by creating state-machine, this is library code.

But I'd glad to use await here, but it's not supported in .Net 4.0 under VS2012 (As far as I know)